application/classes/Ushahidi/Formatter/Post/CSV.php

```diff
@@ -37,113 +37,203 @@ public function __invoke($records)
 
-		// ... else assign value as single string
-		else
-		{
-			$record[$key] = $value;
+	/**
+	 * @param $fields: an array with the form: ["key": (value)] where value can be anything that the user chose.
+	 * @return array of sorted fields with a zero based index. Multivalue keys have the format keyxyz.index index being an arbitrary count of the amount of fields.
+	 */
+	private function createSortedHeading($fields){
+		$headingResult = [];
+		// fieldsWithPriorityValue: an associative array with the form ["uuid"=>[label: string, priority: number, stage: number],"uuid"=>[label: string, priority: number, stage: number]]
+		$fieldsWithPriorityValue = [];
+		/**
+		 * Separate by fields that have custom priority and fields that do not have custom priority assigned
+		 */
+		foreach ($fields as $fieldKey => $fieldAttr) {
+			if (!is_array($fieldAttr)) {
+				$headingResult[$fieldKey] = $fieldAttr;
+			} else if (is_array($fieldAttr) && isset($fieldAttr['nativeField'])){
+				if ($fieldAttr['count'] === 0) {
+					$headingResult[$fieldKey] = $fieldAttr['label'];
+				}
+				for ($i = 0 ; $i < $fieldAttr['count']; $i++){
+					$headingResult[$fieldKey.'.'.$i] = $fieldAttr['label'].'.'.$i;
+				}
+			} else {
+				$fieldsWithPriorityValue[$fieldKey] = $fieldAttr;
+			}
 		}
+		/**
+		 * Sort the non custom priority fields alphabetically, ASC (default)
+		 */
+		uasort($headingResult, function($item1, $item2){
+			return strcmp($item1, $item2);
+		});
+		/**
+		 * sorting the multidimensional array of properties
+		 */
+		/**
+		 * First, group fields by stage and survey id
+		 */
+		$attributeKeysWithStage = $this->groupFieldsByStage($fieldsWithPriorityValue);
+		/**
+		 * After we have group by stage , we can proceed to sort each field by priority inside the stage
+		 */
+		$attributeKeysWithStageFlat = $this->sortGroupedFieldsByPriority($attributeKeysWithStage);
+		/**
+		 * Add the custom priority fields to the heading array and return it, as is.
+		 */
+		$headingResult += $attributeKeysWithStageFlat;
+		return $headingResult;
 	}
 
-	private function assignColumnHeading(&$columns, $key, $label, $value)
-	{
-		if (is_array($value))
-		{
-			// Assign in multiple columns
-			foreach ($value as $sub_key => $sub_value)
-			{
-				$multivalue_key = $key.'.'.$sub_key;
-
-				if (! in_array($multivalue_key, $columns))
-				{
-					$columns[$multivalue_key] = $label.'.'.$sub_key;
+	/**
+	 * @param $groupedFields is an associative array with fields grouped in arrays by their stage
+	 * @return array . Flat, associative. Example => ['keyxyz'=>'label for key', 'keyxyz2'=>'label for key2']
+	 */
+	private function sortGroupedFieldsByPriority($groupedFields){
+		$attributeKeysWithStageFlat = [];
+		foreach ($groupedFields as $stageKey => $attributeKeys){
+			/**
+			 * uasort is used here to preserve the associative array keys when they are sorted
+			 */
+			uasort($attributeKeys, function ($item1, $item2) {
+				if ($item1['priority'] == $item2['priority']){
```

Comment:
I think we can use strict === here since we expect the type to be the same.

Reply:
yeah makes sense.